src/ariel-os-debug/src/lib.rs

```diff
@@ -113,6 +103,59 @@ mod backend {
     }
 }
 
+#[cfg(all(feature = "debug-console", feature = "uart"))]
+#[doc(hidden)]
+pub mod backend {
+    use embassy_sync::once_lock::OnceLock;
+
+    #[doc(hidden)]
+    pub enum Error {
+        Writing,
+    }
+
+    // Populated by a downstream crate.
+    // The function must print to a UART output.
+    #[doc(hidden)]
+    pub static DEBUG_UART_WRITE_FN: OnceLock<fn(&[u8]) -> Result<(), Error>> = OnceLock::new();
+
+    struct DebugUart;
+
+    impl core::fmt::Write for DebugUart {
+        fn write_str(&mut self, s: &str) -> core::fmt::Result {
+            let bytes = s.as_bytes();
+
+            if let Some(debug_uart_write_fn) = DEBUG_UART_WRITE_FN.try_get() {
```

Comment:
This we'll need to revisit. It basically means that ISR output will be lost if a thread is already printing, which is likely to happen given uart writes with only 1 character per ms. We'll need to go ring-buffer, make it line based (so output won't mingle), ...